tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -304,6 +336,80 @@ private GraphOperation graphOp(Operand<?> operand) {
     return seen;
   }
 
+  /**
+   * Get all ops directly or indirectly required to calculate {@code outputs} (not including {@code outputs}), including
+   * control dependencies.
+   *
+   * @param outputs the starting points of the traversal.
+   * @return the ops needed to calculate {@code outputs}, not including {@code outputs}
+   */
+  public synchronized Set<GraphOperation> upstreamOps(Set<GraphOperation> outputs) {
```

Comment:
I missed this earlier, but why are these methods `synchronized`? It seems like they don't mutate the state of the Graph.

Reply:
It doesn't, but if you mutate the graph while they are running the results could be incorrect.  More for downstream (with consumers) though, like if you added a new op using one of the inputs that had already been processed.  I can remove that from upstream.

Reply:
Ok.